### PR TITLE
port tswap buy tests to mpl-core

### PR DESCRIPTION
**Pull request by samuelvanderwaal**

### CODE REVIEW

File: clients/js/test/T22/deposit.test.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  CurveType,
+  fetchPool,
+  findNftDepositReceiptPda,
+  getDepositNftT22InstructionAsync,
+  PoolConfig,
+  PoolType,
+} from '../../src/index.js';
+import { createPool, createWhitelistV2 } from '../_common.js';
+
+test('deposit non-whitelisted NFT fails', async (t) => {
```
Comment (leantOnSol) [inline clients/js/test/T22/deposit.test.ts:30]: Maybe 2 additional tests could fit in here:

1. it can deposit a whitelisted NFT
2. it can't deposit a whitelisted NFT with incorrect rem accs (transferHookAccs)

File: clients/js/test/depositSol.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import test from 'ava';
+import {
+  PoolType,
+  TENSOR_AMM_ERROR__WRONG_POOL_TYPE,
+  getDepositSolInstruction,
+} from '../src/index.js';
+import { ONE_SOL, TestAction } from './_common.js';
+import { setupLegacyTest } from './legacy/_common.js';
+
+test('deposit SOL into a NFT pool fails', async (t) => {
```
Comment (leantOnSol) [inline clients/js/test/depositSol.test.ts:17]: similarly here, shouldnt we double check here that depositing SOL into a Token/Trade pool succeeds?

File: clients/js/test/editPool.test.ts [hunk 3]
```diff
@@ -15,8 +23,101 @@ import {
...
+  TestAction,
+  createPool,
+  createWhitelistV2,
+  nftPoolConfig,
+  tokenPoolConfig,
+  tradePoolConfig,
+} from './_common.js';
+import { setupLegacyTest, testSell } from './legacy/_common.js';
+
+test('can edit pool with new config', async (t) => {
```
Comment (leantOnSol) [inline clients/js/test/editPool.test.ts:53]: 2 pretty important tests missing here:

1. it can overwrite makerBroker on edit
2. it can choose to not overwrite brokers - prior makerBroker should stay

File: clients/js/test/legacy/buy.test.ts [hunk 4]
```diff
@@ -218,11 +261,34 @@ test('buy from NFT pool, skip non-rent-exempt creators', async (t) => {
...
+
+  await testBuy(t, legacyTest, {
     brokerPayments: false,
-    creators,
+    creators, // pass in the verified creators
+    checkCreatorBalances: false,
+    pNft: true,
   });
 
+  // First three creators should have a higher balance.
```
Comment (leantOnSol) [inline clients/js/test/legacy/buy.test.ts:286]: why are we not checking exact creator balance diffs depending on their share?

File: clients/js/test/legacy/sell.test.ts [hunk 5]
```diff
@@ -1510,3 +2480,304 @@ test('token pool with makerBroker set requires passing the account in & fails w/
   // Should fail with a missing makerBroker error.
   await expectCustomError(t, promise, TENSOR_AMM_ERROR__WRONG_MAKER_BROKER);
 });
+
+test('alternate deposits & sells', async (t) => {
```
Comment (leantOnSol) [inline clients/js/test/legacy/sell.test.ts:2484]: what program functionality is this test supposed to assert?

File: clients/js/test/mpl_core/buy.test.ts [hunk 6]
```diff
@@ -11,178 +11,535 @@ import {
...
+    await createDefaultTransaction(client, traderB),
+    (tx) => appendTransactionMessageInstruction(buyNftIx, tx),
+    (tx) => signAndSendTransaction(client, tx)
+  );
 
-  // Deposit Receipt is closed
-  await assertNftReceiptClosed({ t, client, pool, mint: asset.address });
+  await expectCustomError(t, promise, ANCHOR_ERROR__CONSTRAINT_SEEDS);
+
+  // Try to buy NFT A from pool B, the default derived NFT receipt will be from
```
Comment (leantOnSol) [inline clients/js/test/mpl_core/buy.test.ts:401]: One regression test idea for all buys/sells:

We could abstract the IdlBuffer test flow + try to imitate deposit receipts to assert discriminator checks

File: clients/js/test/mpl_core/deposit.test.ts [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+  CurveType,
+  fetchPool,
+  findNftDepositReceiptPda,
+  getDepositNftCoreInstructionAsync,
+  PoolConfig,
+  PoolType,
+} from '../../src/index.js';
+import { createPool, createWhitelistV2 } from '../_common.js';
+
+test('deposit non-whitelisted NFT fails', async (t) => {
```
Comment (leantOnSol) [inline clients/js/test/mpl_core/deposit.test.ts:35]: similar to prior deposit files: first test should probably be "deposit whitelisted NFT succeeds"

File: clients/js/test/mpl_core/withdraw.test.ts [hunk 8]
```diff
@@ -5,6 +5,13 @@ import {
   appendTransactionMessageInstruction,
   pipe,
 } from '@solana/web3.js';
+import {
```
Comment (leantOnSol) [inline clients/js/test/mpl_core/withdraw.test.ts:8]: missing test variants for NFT pools (both "it can withdraw an NFT from an NFT pool" and "it cannot withdraw an NFT from an NFT pool with wrong owner")

File: clients/js/test/withdrawSol.test.ts [hunk 9]
```diff
@@ -13,15 +14,16 @@ import test from 'ava';
 import {
   PoolType,
   TENSOR_AMM_ERROR__POOL_INSUFFICIENT_RENT,
+  TENSOR_AMM_ERROR__WRONG_POOL_TYPE,
```
Comment (leantOnSol) [inline clients/js/test/withdrawSol.test.ts:17]: Only trade pool tests, Token pool test variants missing entirely

File: clients/js/test/depositSol.test.ts (hunk 2 above)
Comment (samuelvanderwaal) [reply clients/js/test/depositSol.test.ts:17]: We could do that explicitly, but `setupLegacyTest` deposits SOL for all the sell tests and those test both pool types.

File: clients/js/test/legacy/sell.test.ts (hunk 5 above)
Comment (samuelvanderwaal) [reply clients/js/test/legacy/sell.test.ts:2484]: This is ported over from TSWAP, it's just doing a bunch of deposits and sells to make sure pools don't have edge cases for a lot of actions.

File: clients/js/test/T22/deposit.test.ts (hunk 1 above)
Comment (samuelvanderwaal) [reply clients/js/test/T22/deposit.test.ts:30]: Added

File: clients/js/test/mpl_core/deposit.test.ts (hunk 7 above)
Comment (samuelvanderwaal) [reply clients/js/test/mpl_core/deposit.test.ts:35]: Added

File: clients/js/test/editPool.test.ts (hunk 3 above)
Comment (samuelvanderwaal) [reply clients/js/test/editPool.test.ts:53]: The contract doesn't currently allow overwriting the `maker_broker` on edit, not sure if it should. @richardwu any reason not to allow this?

File: clients/js/test/legacy/buy.test.ts (hunk 4 above)
Comment (samuelvanderwaal) [reply clients/js/test/legacy/buy.test.ts:286]: Test fatigue. Fixed.

File: clients/js/test/mpl_core/withdraw.test.ts (hunk 8 above)
Comment (samuelvanderwaal) [reply clients/js/test/mpl_core/withdraw.test.ts:8]: Added

File: clients/js/test/withdrawSol.test.ts (hunk 9 above)
Comment (samuelvanderwaal) [reply clients/js/test/withdrawSol.test.ts:17]: Added

File: clients/js/test/editPool.test.ts (hunk 3 above)
Comment (richardwu) [reply clients/js/test/editPool.test.ts:53]: Seems weird to be able to edit it after you make the order, since it's the originator that should get the credit.

Think we should force them to close/re-open if they really want to change the maker broker.

Comment (leantOnSol) [reply clients/js/test/editPool.test.ts:53]: Hmmmmmmm, not sure I agree - this would be very weird, because for marketplace bids we do allow the change of `makerBroker` on bid edits.

I feel like we should be consistent here or that'll cause confusion @richardwu

Comment (leantOnSol) [reply clients/js/test/editPool.test.ts:53]: https://github.com/tensor-foundation/marketplace/blob/main/program/src/instructions/bid.rs#L79

there is no check that this is only getting called when initialized (e.g. rentPayer)

Comment (leantOnSol) [reply clients/js/test/editPool.test.ts:53]: same with edits on listings (https://github.com/tensor-foundation/marketplace/blob/main/program/src/instructions/edit.rs#L31)

Feels very off to not allow the edit of `makerBroker` on pools then

File: clients/js/test/mpl_core/buy.test.ts (hunk 6 above)
Comment (samuelvanderwaal) [reply clients/js/test/mpl_core/buy.test.ts:401]: What's the advantage to doing this in addition to the fake pool tests with idlBuffer?

File: program/src/instructions/shared_accounts.rs [hunk 10]
```diff
@@ -114,6 +114,7 @@ pub struct TradeShared<'info> {
             pool.pool_id.as_ref(),
         ],
         bump = pool.bump[0],
+        has_one = owner @ ErrorCode::WrongOwner,
```
Comment (richardwu) [inline program/src/instructions/shared_accounts.rs:117]: Hmmm I removed this b/c it was redundant with the seeds.

otherwise to be consistent we'd have to add this to all the ixs again...

Comment (samuelvanderwaal) [reply program/src/instructions/shared_accounts.rs]: I think a has_one constraint is better than a seeds here if we're only going to have one, but for consistency I removed it.

**Review by richardwu — APPROVED** [review]
